plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java

```diff
@@ -0,0 +1,372 @@
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class OpenLineageListener
+        implements EventListener
+{
+    private static class StringTokenProvider
+            implements TokenProvider
+    {
+        private final String token;
+
+        public StringTokenProvider(String token)
+        {
+            this.token = token;
+        }
+
+        @Override
+        public String getToken()
+        {
+            return this.token;
+        }
+    }
+
+    private final OpenLineage ol = new OpenLineage(URI.create("https://github.com/trinodb/trino/plugin/trino-openlineage"));
+    private final OpenLineageClient client;
+    private final String jobNamespace;
+    private final List<QueryType> includeQueryTypes;
+    private static final Logger logger = Logger.get(OpenLineageListener.class);
+
+    @Inject
+    public OpenLineageListener(OpenLineageListenerConfig listenerConfig, Optional<OpenLineageClientHttpTransportConfig> httpTransportConfig)
+            throws Exception
+    {
+        this.client = getClient(listenerConfig, httpTransportConfig);
+
+        this.jobNamespace = String.format("trino://%s:%s", new URI(listenerConfig.getTrinoHost()).getHost(), listenerConfig.getTrinoPort());
+
+        this.includeQueryTypes = listenerConfig.getIncludeQueryTypes();
+
+        logger.info("Supported query types" + this.includeQueryTypes);
+    }
+
+    private OpenLineageClient getClient(OpenLineageListenerConfig listenerConfig, Optional<OpenLineageClientHttpTransportConfig> httpTransportConfig)
+            throws Exception
+    {
+        OpenLineageClient.Builder clientBuilder = OpenLineageClient.builder();
+
+        if (listenerConfig.getTransport().equals(OpenLineageTransport.HTTP) && httpTransportConfig.isPresent()) {
+            logger.info("Using HttpTransport");
+            OpenLineageClientHttpTransportConfig config = httpTransportConfig.get();
+            clientBuilder.transport(
+                    new HttpTransport(
+                            new HttpConfig(
+                                    new URI(config.getUrl()),
+                                    config.getEndpoint(),
+                                    null,
+                                    config.getTimeoutInMillis(),
+                                    config.getApiKey().isPresent() ? new StringTokenProvider(config.getApiKey().get()) : null,
+                                    config.getUrlParams(),
+                                    config.getHeaders())));
+        }
+        else if (listenerConfig.getTransport() == OpenLineageTransport.CONSOLE) {
+            logger.info("Using ConsoleTransport");
+            clientBuilder.transport(new ConsoleTransport());
+        }
```

Comment:
We can extract some sort of an interface aka Decorator which would apply these changes on top `OpenLineageClient.Builder` and each authentication mechansim can have its own Decorator which would be binded as a part of its module.